### PR TITLE
Add option to define custom validation rules

### PR DESCRIPTION
**Pull request by mic-kul**
This PR adds ability to define custom validation rule using `Proc`:

```
config.middleware.use RackPassword::Block, auth_codes: ['janusz'], custom_rule: proc { |request| request.env['HTTP_USER_AGENT'].include?('facebook') }
```

### CODE REVIEW

File: lib/rack_password.rb [hunk 1]
```diff
@@ -64,6 +64,11 @@ def valid_code? code
       return false if @options[:auth_codes].nil?
       @options[:auth_codes].include? code
     end
+
+    def custom_rule?
```
Comment (kamil89) [inline lib/rack_password.rb]: what about to use `valid_custom_rule?` name?

File: spec/lib/rack_password/block_validator_spec.rb [hunk 2]
```diff
@@ -58,4 +58,27 @@
       expect(bv.valid_code?("incorrect_secret")).to be(false)
     end
   end
+
+  describe "proc control" do
+    context 'with proc allowing to pass' do
```
Comment (kamil89) [inline spec/lib/rack_password/block_validator_spec.rb]: single or double quotes?

File: spec/lib/rack_password/block_validator_spec.rb [hunk 3]
```diff
@@ -58,4 +58,27 @@
...
+      let(:options) { Hash[auth_codes: ["secret"], key: :staging_auth, custom_rule: proc { true } ] }
+      let(:request) { double "Request" }
+
+      it "be true when proc evaluates to true" do
```
Comment (kamil89) [inline spec/lib/rack_password/block_validator_spec.rb]: `it "is true..."`

**Comment by kamil89** [comment]
I'd add one more test case: check if `valid?` behaves correctly when `valid_custom_rule?` is true and rest conditions are false.

Overall, very useful feature!

File: spec/lib/rack_password/block_validator_spec.rb [hunk 4]
```diff
@@ -58,4 +58,27 @@
...
+        bv = RackPassword::BlockValidator.new(options, request)
+        expect(bv.custom_rule?).to be(true)
+      end
+    end
+
+    context 'with proc set to deny-all' do
+      let(:options) { Hash[auth_codes: ["secret"], key: :staging_auth, custom_rule: proc { false } ] }
+      let(:request) { double "Request", path: '/', ip: "127.0.0.1", cookies: { } }
+
+      it "be true when proc evaluates to true" do
```
Comment (kamil89) [inline spec/lib/rack_password/block_validator_spec.rb]: as above `"is true..."`

**Comment by kamil89** [comment]
Looks good for me.

**Comment by mic-kul** [comment]
Thanks @kamil89

**Pull request merged**